### DISCUSSION TITLE
Suggestion needed.

### DISCUSSION THREAD

**Discussion by fahmidul** [abertschi/water-me#28] [General]
I was trying to make a water reminder app. Currently, I am a little stuck on a feature where the user will be able to set a periodic reminder after every x (i.e 15 minutes) times except their selected sleeping time. The problem is when I was using a work manager for scheduling that reminder job it's firing reminder events at various time duration (i.e: after 30 minutes).
Did you ever experience something like that? Or any suggestions on how can I overcome this issue.
Thanks.

**Comment by abertschi** [comment]
Periodic task are subject to Android battery saving policies, you may not get accurate reminders with work manager. You can overcome the issue by scheduling a task with alarm manger. This is how the alarm clock is implemented. However, battery performance will suffer and it's typically not best practice. If you stick to work manger you can implement a callback which keeps track of the last reminder and will issue the notification once enough time has passed and block earlier invocations.

Alarm manager: https://github.com/abertschi/syncopoli/blob/feature/sync/app/src/main/java/org/amoradi/syncopoli/ScheduleManager.java#L55

**Reply by abertschi** [reply]
Alternatively, you may remove your app from doze restrictions in app settings (battery usage/ battery optimization/ select not optimized)

**Reply by abertschi** [reply]
Work manager has caused troubles by too restrictive OS scheduling of some device vendors in the past.

https://issuetracker.google.com/issues/113676489?pli=1  
https://stackoverflow.com/questions/68003143/unique-periodic-work-manager-not-working-in-xiaomi-phone  
https://issuetracker.google.com/issues/113676489?pli=1

**Reply by fahmidul** [reply]
So kind of you. I am grateful for your quick response.
Do you have any Idea how the default alarm clock work? Is it based on the alarm manager? 
Is it possible to make an alarm app that will perform exactly like the default alarm app? I am especially talking about the one feature of an alarm app that If the user misses the alarm then fire the alarm after a specific time again and again...

**Reply by abertschi** [reply]
In the first link i posted in ScheduleManager.java is an example, you register an intent with AlarmManager and a frequency. The alarm manager service will then invoke your intent. In case you target flutter and not Android you may have to implement wrappers for dart.

**Reply by abertschi** [reply]
There is also this: https://f-droid.org/en/packages/io.github.z3r0c00l_2k.aquadroid/ Which you may want to fork and extend.

**Reply by abertschi** [reply]
> So kind of you. I am grateful for your quick response. Do you have any Idea how the default alarm clock work? Is it based on the alarm manager? 

The stock android alarm clock is implemented with alarm manager:

https://android.googlesource.com/platform/packages/apps/DeskClock/+/refs/heads/master/src/com/android/deskclock/alarms/AlarmStateManager.kt#274

**Reply by fahmidul** [reply]
> There is also this: https://f-droid.org/en/packages/io.github.z3r0c00l_2k.aquadroid/ Which you may want to fork and extend.

I have run this project earlier, Unfortunately, the Reminder firing function doesn't work properly.

**Comment by fahmidul** [comment]
Thanks a lot for your suggestions and resources.

**Reply by abertschi** [reply]
Sure thing, if you want to contribute to water-me, we have have open issues, some are quite easy to implement ;)

https://github.com/abertschi/water-me/issues

cheers

**Reply by fahmidul** [reply]
I am interested and would be happy If I can contribute a little. Please let me understand this project structure first. I will be back here soon.

**Reply by abertschi** [reply]
sure thing, let me know if you need help. I can pin point you to some places if things are unclear.

**Reply by fahmidul** [reply]
Sure, I will let you know soon since I am on a festival vacation.
